Add batch support

ziadsawalha opened the pull request:
Giving this a stab. Feedback welcome!

pcreux commented on lib/csv_importer/csv_reader.rb:
```diff
@@ -25,12 +34,23 @@ def csv_rows
+
+      if use_batches
+        io = file_or_content_io
+        separator = detect_separator(io.read(1024))
```
We should extract `separator` to a method and memoize it as it's called twice.

pcreux commented on lib/csv_importer/runner.rb:
```diff
@@ -12,11 +12,26 @@ def self.call(*args)
+        batch_size: batch_size,
+        use_batches: use_batches
+      )
+    end
```
Instead of overwriting `initialize`, what do you think about extracting `csv_reader` to a method and memoizing it?

pcreux commented on lib/csv_importer/csv_reader.rb:
```diff
@@ -47,13 +67,52 @@ def read_content
+      end
+    end
+
+    def sanitize_content_stream(io)
```
Great method but it's not used. Can you make the `Enumerator` use it?

pcreux commented:
This is great @ziadsawalha ! I added a couple of comments.

Could creating a new class simplify the codebase? e.g. `Runner` could toggle between `CSVBulkReader` and `CSVBatchReader` depending on `use_batches`.

Regarding the public API, `use_batches` doesn't sound great IMO. What about `batch_load`?

We could release this as the next `minor` release and I would like to release a `major` version where we remove the `virtus` dependency. Thoughts?

ziadsawalha commented:
Thanks for the good feedback. I'll submit updates when I get to it... hopefully tonight (Europe).

ziadsawalha commented:
I've updated the PR with the following changes:
- Extracted bulk/streaming code to a `CSVBatchReader` which inherits from `CSVReader` (I opted not to rename `CSVReader` ... unless you insist!)
- Extended `Config` and `Runner` to support enabling batch loading
- Renamed flag to `batch_load`
- Added some tests (I think more are needed)
- Minor edits to files I touched

I struggled with trying to support streams that don't have rewind (which means we need to detect the separator and headers on load and then not try to restart the stream). That complicated things a bit, IMO.
 
More feedback welcome while I try to add more tests and validate this in real life use cases.

Commit:
Add batch support

pcreux commented:
Great, thank you for making these changes @ziadsawalha !

How does this work against real-life use cases?

What kind of streams don't have rewind? S3? 

Could we close and re-open (or duplicate?!) the stream if it doesn't support rewinding?
